File: tree/ordered_traversal/144_binary_tree_preorder_traversal.cc
#include <iostream>
#include <stack>
#include <vector>

struct TreeNode {
  int value;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : value(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : value(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : value(x), left(left), right(right) {}
};

void Preorder(TreeNode* root, std::vector<int>& result) {
  if (root == nullptr) {
    return;
  }
  result.push_back(root->value);
  Preorder(root->left, result);
  Preorder(root->right, result);
}

std::vector<int> PreorderTraversal(TreeNode* root) {
  std::vector<int> result;
  Preorder(root, result);
  return result;
}

std::vector<int> PreorderTraversalIteration(TreeNode* root) {
  std::vector<int> result;
  std::stack<TreeNode*> stk;
  while (root || !stk.empty()) {
    while (root) {
      result.push_back(root->value);
      stk.push(root);
      root = root->left;
    }
    root = stk.top();
    stk.pop();
    root = root->right;
  }
  return result;
}

std::vector<int> PreorderTraversalIteration2(TreeNode* root) {
  std::vector<int> result;
  if (!root) {
    return result;
  }
  std::stack<TreeNode*> stack;
  stack.push(root);
  while (!stack.empty()) {
    TreeNode* node = stack.top();
    stack.pop();
    result.push_back(node->value);
    if (node->right) {
      stack.push(node->right);
    }
    if (node->left) {
      stack.push(node->left);
    }
  }
  return result;
}

int main(int argc, char* argv[]) {
  TreeNode* node_1 = new TreeNode(1);
  TreeNode* node_2 = new TreeNode(2);
  TreeNode* node_3 = new TreeNode(3);
  node_1->right = node_2;
  node_2->left = node_3;
  std::vector<int> result = PreorderTraversal(node_1);
  for (int value : result) {
    std::cout << value << '\t';
  }
  std::cout << std::endl;
  delete node_1;
  delete node_2;
  delete node_3;
  return 0;
}

File: tree/ordered_traversal/144_binary_tree_preorder_traversal.py
from typing import List, Optional
from tree.tree_node import TreeNode, create_tree_node, print_tree_node


# 2026/05/19: 自己实现递归和迭代版本
def preorder(root: Optional[TreeNode], res: List[int]) -> None:
    if root is None:
        return
    res.append(root.val)
    preorder(root.left, res)
    preorder(root.right, res)


def preorder_traversal1(root: Optional[TreeNode]) -> List[int]:
    res = []
    preorder(root, res)
    return res


def preorder_traversal2(root: Optional[TreeNode]) -> List[int]:
    res = []
    stack = []

    if root is None:
        return res
    stack.append(root)
    while len(stack) > 0:
        curr = stack.pop()
        res.append(curr.val)
        if curr.right is not None:
            stack.append(curr.right)
        if curr.left is not None:
            stack.append(curr.left)

    return res


root = [1, 2, 3, 4, 5, None, 8, None, None, 6, 7, 9]
root = create_tree_node(root)
print_tree_node(root)
print(preorder_traversal2(root))
